File: app/src/main/java/com/emiryanvl/tapreader/domain/repositories/BookRepository.java
package com.emiryanvl.tapreader.domain.repositories;

import com.emiryanvl.tapreader.domain.model.Book;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Observable;

public interface BookRepository {
    Observable<List<Book>> getAllBooks();

    Observable<Book> getBook(int id);

    Completable addBook(Book book);

    void updateBook(int id, Book book);

    void deleteBook(int id);
}

File: app/src/main/java/com/emiryanvl/tapreader/domain/repositories/BookRepository.kt
package com.emiryanvl.tapreader.domain.repositories

import com.emiryanvl.tapreader.domain.models.Book
import kotlinx.coroutines.flow.Flow

interface BookRepository {

    fun getAllBooks(): Flow<List<Book>>

    fun getBooksByQuery(query: String): Flow<List<Book>>

    fun getBooksBySubject(subject: String): Flow<List<Book>>

    fun getBook(id: Int): Flow<Book>

    suspend fun addBook(book: Book)

    suspend fun updateBook(id: Int, book: Book)

    suspend fun deleteBook(id: Int)
}